### PR TITLE
Use generators instead of lists

### PR DESCRIPTION
**Pull request by kaapstorm**
Context at [FB 185783](http://manage.dimagi.com/default.asp?185783).

Instead of building lists of cases, use a generator when fetching cases, and a generator for the cases to be updated.

I've kept most of this code untouched. Hold off on merge for a bit. I'd like to PEP8 a little, and see if I can use [IterDB](https://github.com/dimagi/commcare-hq/blob/master/corehq/util/couch.py#L103-103) for the updates.

@snopoke, @sravfeyn, cc @millerdev

### CODE REVIEW

File: custom/_legacy/hsph/tasks.py [hunk 1]
```diff
@@ -235,22 +220,27 @@ def new_update_case_properties():
...
+            'create': False,
+            'case_id': case['case_id'],
+            'update': case['update'],
+            'close': case['close'],
+        }
+        if case.get('owner_id', None):
+            kwargs['owner_id'] = case['owner_id']
+        case_block = ElementTree.tostring(CaseBlock(**kwargs).as_xml())
+        # FB 185783: Do one at a time
+        submit_case_blocks(case_block, domain)
```
Comment (kaapstorm) [inline custom/_legacy/hsph/tasks.py]: I'm curious whether it's actually necessary to do this bit one at a time, considering we're already using generators to fetch the cases, and to return those that need to be modified.

File: custom/_legacy/hsph/tasks.py [hunk 2]
```diff
@@ -95,10 +92,8 @@ def new_update_case_properties():
     past_42_date = past_x_date(time_zone, 42)
     setup_indices()
     for domain in DOMAINS:
-        case_list = list(get_cases_in_domain(domain, type=BIRTH_TYPE))
-        case_list = case_list + list(get_cases_in_domain(domain, type=CATI_FIDA_CHECK_TYPE))
-        cases_to_modify = []
-        for case in case_list:
+        for case in chain(get_cases_in_domain(domain, type=BIRTH_TYPE),
+                          get_cases_in_domain(domain, type=CATI_FIDA_CHECK_TYPE)):
```
Comment (snopoke) [inline custom/_legacy/hsph/tasks.py]: No point in fetching the whole case if you're just returning the ID. Use `get_case_ids_in_domain` instead

File: custom/_legacy/hsph/tasks.py (hunk 1 above)
Comment (snopoke) [inline custom/_legacy/hsph/tasks.py]: the case importer does them in batches of 100

**Comment by snopoke** [comment]
Looking at this again I don't think you can use `IterDb` since you need to submit case blocks and not modify the case object directly.

File: custom/_legacy/hsph/tasks.py [hunk 3]
```diff
@@ -1,32 +1,31 @@
...
 from corehq.apps.groups.models import Group
 from corehq.apps.hqcase.dbaccessors import get_cases_in_domain
 from corehq.apps.hqcase.utils import submit_case_blocks
 from dimagi.utils.decorators.memoized import memoized
 
+
 DOMAINS = ["hsph-dev", "hsph-betterbirth", "hsph-learning-sites", "hsph-test"]
 PAST_N_DAYS = 21
-GROUPS_TO_CHECK = ["cati", "cati-tl"]
-GROUP_SHOULD_BE = "fida"
```
Comment (kaapstorm) [inline custom/_legacy/hsph/tasks.py]: (Unused anywhere in commcare-hq)

File: custom/_legacy/hsph/tasks.py [hunk 4]
```diff
@@ -78,15 +78,11 @@ def get_group_id(domain, owner_type, facility_id):
...
+past_x_date = lambda tz, past_x_days: (datetime.datetime.now(tz) - datetime.timedelta(past_x_days)).date()
+get_none_or_value = lambda _obj, _attr: getattr(_obj, _attr) if (hasattr(_obj, _attr)) else ''
 
 
-@periodic_task(
-    run_every=crontab(minute=1, hour="*/6"),
-    queue=getattr(settings, 'CELERY_PERIODIC_QUEUE', 'celery')
-)
-def new_update_case_properties():
+def get_cases_to_modify():
```
Comment (millerdev) [inline custom/_legacy/hsph/tasks.py]: Would it make sense to rename this to `iter_cases_to_modify`? Either that or add a docstring that explains that it returns an generator, not a list. The size of the function makes it hard to tell by glancing at the code.

File: custom/_legacy/hsph/tasks.py [hunk 5]
```diff
@@ -78,15 +78,11 @@ def get_group_id(domain, owner_type, facility_id):
...
-get_none_or_value = lambda _object, _attribute: getattr(_object, _attribute) if (hasattr(_object, _attribute)) else ''
+past_x_date = lambda tz, past_x_days: (datetime.datetime.now(tz) - datetime.timedelta(past_x_days)).date()
+get_none_or_value = lambda _obj, _attr: getattr(_obj, _attr) if (hasattr(_obj, _attr)) else ''
 
 
-@periodic_task(
-    run_every=crontab(minute=1, hour="*/6"),
-    queue=getattr(settings, 'CELERY_PERIODIC_QUEUE', 'celery')
-)
-def new_update_case_properties():
```
Comment (millerdev) [inline custom/_legacy/hsph/tasks.py]: Was there a reason this was moved to the bottom of the file? It makes the diff particularly hard to read.

**Comment by kaapstorm** [comment]
Good points @millerdev. Fixed.

**Comment by millerdev** [comment]
:+1:

File: custom/_legacy/hsph/tasks.py (hunk 2 above)
Comment (sravfeyn) [inline custom/_legacy/hsph/tasks.py]: We do need whole case info

**Comment by sravfeyn** [comment]
Thanks for the cleanup!

**Pull request merged**